rfc/XXX-trusted-kubernetes-images.md

```diff
@@ -0,0 +1,126 @@
+part of Constellation's TCB, but it is also a problem when Constellation is set
+up in a restricted environment where this repo is not available.
+
+## Requirements
+
+1. In a default installation, Constellation must verify Kubernetes control plane images.
+2. Users must be able to override the image repository for Kubernetes control plane images.
+3. Users must be able to override the verification mechanism for Kubernetes control plane images.
+
+Out of scope:
+
+- reproducibility from github.com/kubernetes/kubernetes to registry.k8s.io and
+  the associated chain of trust
+- container registry trust & CA certificates
+
+## Solution
+
+Kubernetes control plane images are going to be pinned by hash and verified by
+the CRI. In a default installation, we take the upstream images from
+registry.k8s.io and ship hashes with the CLI. Users can override the image
+per control plane component.
+
+### Image Hashes
+
+We are concerned with the following control plane images (for v1.27.7):
+
+- registry.k8s.io/kube-apiserver:v1.27.7
+- registry.k8s.io/kube-controller-manager:v1.27.7
+- registry.k8s.io/kube-scheduler:v1.27.7
+- registry.k8s.io/coredns/coredns:v1.10.1
+- registry.k8s.io/etcd:3.5.9-0
+
+When a new Kubernetes version is added to `/internal/versions/versions.go`, we
+generate the corresponding list of images with `kubeadm config images list` and
+probe their hashes on registry.k8s.io. Generating the list of images this way
+must happen offline to prevent `kubeadm` from being clever. These hashes are
+added to `versions.go` as a mapping from component to pinned image:
+
+```golang
+V1_27: {
+  ClusterVersion: "v1.27.7",
+  // ...
+  ControlPlaneImages: {
+    "kube-apiserver": "registry.k8s.io/kube-apiserver:v1.27.7@sha256:<...>", 
+    "kube-controller-manager": "registry.k8s.io/kube-controller-manager:v1.27.7@sha256:<...>", 
+    "kube-scheduler": "registry.k8s.io/kube-scheduler:v1.27.7@sha256:<...>", 
+    "etcd": "registry.k8s.io/etcd:3.5.9-0@sha256:<...>",
+    "coredns": "registry.k8s.io/coredns/coredns:v1.10.1@sha256:<...>",
+  },
+}
+```
+
+### Configuration Options
+
+Users can override the images in their `constellation-conf.yaml`, using the
+same nested structure as `versions.go` (*to be elaborated on*).
+
+### Cluster Init
+
+During cluster initialization, we need to tell `kubeadm` that we want to use
+the embedded image references instead of the default ones. For that, we
+populate the
+[`InitConfiguration.Patches`](https://pkg.go.dev/k8s.io/kubernetes@v1.27.7/cmd/kubeadm/app/apis/kubeadm/v1beta3#InitConfiguration)
+with a list of patch files that replace the container image with the pinned
+alternative.
+
+This method does not work for coredns, which can only be customized with the
+[`ClusterConfiguration.DNS`](https://pkg.go.dev/k8s.io/kubernetes@v1.27.7/cmd/kubeadm/app/apis/kubeadm/v1beta3#ClusterConfiguration)
+section. It's a bit of a hack, but we can parse the registry part and the
```

Comment:
The hack part is that it would look like you could do

```yaml
"coredns": "my.custom.registry/my-custom-coredns:123@sha256:..."
```

but this can't be modeled with `ClusterConfiguration.DNS` because the image path is fixed.

In the spirit of https://github.com/edgelesssys/constellation/pull/2648/files#r1406263649 it would make sense to manage coredns as a microservice.

Reply:
Adjusted wording.